presto-iceberg/src/main/java/com/facebook/presto/iceberg/optimizer/IcebergPlanOptimizer.java

```diff
@@ -0,0 +1,163 @@
+    }
+
+    private static class FilterPushdownVisitor
+            extends PlanVisitor<PlanNode, Void>
+    {
+        private final IcebergMetadataManager metadataManager;
+        private final ConnectorSession session;
+        private final RowExpressionService rowExpressionService;
+        private final StandardFunctionResolution functionResolution;
+        private final TypeManager typeManager;
+        private final PlanNodeIdAllocator idAllocator;
+
+        public FilterPushdownVisitor(
+                IcebergMetadataManager metadataManager,
+                StandardFunctionResolution functionResolution,
+                RowExpressionService rowExpressionService,
+                TypeManager typeManager,
+                PlanNodeIdAllocator idAllocator,
+                ConnectorSession session)
+        {
+            this.metadataManager = metadataManager;
+            this.functionResolution = functionResolution;
+            this.rowExpressionService = rowExpressionService;
+            this.typeManager = typeManager;
+            this.idAllocator = idAllocator;
+            this.session = session;
+        }
+
+        @Override
+        public PlanNode visitPlan(PlanNode node, Void context)
+        {
+            ImmutableList.Builder<PlanNode> children = ImmutableList.builder();
+            boolean changed = false;
+            for (PlanNode child : node.getSources()) {
+                PlanNode newChild = child.accept(this, null);
+                if (newChild != child) {
+                    changed = true;
+                }
+                children.add(newChild);
+            }
+
+            if (!changed) {
+                return node;
+            }
+            return node.replaceChildren(children.build());
+        }
+
+        @Override
+        public PlanNode visitFilter(FilterNode filter, Void context)
+        {
+            if (!(filter.getSource() instanceof TableScanNode)) {
+                return visitPlan(filter, context);
+            }
+
+            TableScanNode tableScan = (TableScanNode) filter.getSource();
+
+            Map<String, IcebergColumnHandle> nameToColumnHandlesMapping = tableScan.getAssignments().entrySet().stream()
+                    .collect(Collectors.toMap(e -> e.getKey().getName(), e -> (IcebergColumnHandle) e.getValue()));
+
+            RowExpression filterPredicate = filter.getPredicate();
+
+            //TODO we should optimize the filter expression
+            DomainTranslator.ExtractionResult<Subfield> decomposedFilter = rowExpressionService.getDomainTranslator()
+                    .fromPredicate(session, filterPredicate, new SubfieldExtractor(functionResolution, rowExpressionService.getExpressionOptimizer(), session).toColumnExtractor());
+
+            // Only pushdown the range filters which apply to entire columns, because iceberg does not accept the filters on the subfields in nested structures.
```

Comment:
remove period at the end of one line comment? or we might have to add period to other one line comments

Reply:
good call, done